Add SHAP interaction effects, fix minor bug, and add cox loss

slundberg opened the pull request:
This PR shares several local updates I have made:

1. It adds the ability to compute pairwise SHAP interaction effects (using `interaction_contribs=True` in python). This extends the feature contribution explanations so they can now explain the pairwise interactions as well as the marginal feature contributions. This is based on ideas in game theory to extend the Shapley value to higher order interactions (http://www.sciencedirect.com/science/article/pii/S0899825605000278) The result is the ability to provide a matrix of interaction effects for any prediction.

2. There was a bug in the current feature contribution code (noted by Andrew Long) that caused the base rate (the last column in the returned matrix) to be wrong when the gamma was varied during training. This fixes that.

3. A full brute force python version of the explanation code used to better unit test the fast (and more complicated) C++ algorithm.

4. Since I had it in my branch I included a Cox Proportional hazards loss for XGBoost. It uses the regular label vector format and encodes censoring with the sign of the variable. Negative values v are censored at the time abs(v), positive values are events at the given time. It uses the Breslow method for ties.

slundberg commented:
Note this seems to have picked up my previous (already merged) commits on the same fork (5c07a99 and earlier), so if it is important to clear those out let me know.

khotilov commented:
@slundberg Could you please rebase against the master, fix merge conflicts, and do not revert the changes for the spacing around binary operators? That is advised by google c++ styleguide and, frankly, is quite more readable in variety of fonts. 

I'll review the details later. Just a few notes and questions after scanning through the code:
- The results are retuned as 3D tensor (is it symmetric in 2nd vs 3rd dimension?), and it might be very easy to hit memory limits, especially with wide sparse matrices. Even for a 100K rows and 200 features dataset it would result in a ~ 32G output when using doubles. Perhaps it would make sense to have some safeguard check that would prevent evaluation when the expected output size would be above half RAM.
- With interactions, it's usually only the largest ones that matter. Did you consider returning contributions of a limited number of the largest interactions plus "the rest" in a sparse matrix structure? That might be really helpful in those situations when data remains fairly wide even after feature selection.

On Cox PH objective:
- It would make more sense as a separate PR, but ok...
- I agree that hijacking the label sign to use as a censoring indicator is the best poor man's option for doing a simple right-censored Cox regression without much extra work. Unfortunately, that would not be enough for any more involved survival models :(
- Currently, these objective & metric expect the labels to be sorted, which is an awkward requirement to deal with when doing data splits or CV (are you really up for properly modifying the wrappers' code for that?). Another more practical option would be to require the labels to be encoded as discrete time indices, which would allow working with indexed vectors instead of sorting. These methods also could coexist as two parameter-controlled options.
  - Note that for the discrete time option (or in general when large number of ties is expected), having the Efron tie handling method would be preferable.

slundberg commented:
@khotilov thanks for the feedback. I'll go ahead and rebase things on master. My intention when merging was to add the space around binary operators, but it looks like a missed a couple, I can update those.

- It is ncolumns by ncolumns+1 since I left the last column with the base rate in it. For clarity is would probably make sense to keep it symmetric and drop the base rate column. A mem check is a good idea, I'll add that.
- I agree that a sparse matrix format would make sense. To really make that work though we would need to integrate a sparse representation into the original feature attribution code, since the interaction code builds on it.

For cox:
- Yeah it was poor planning getting the changes mixed up during development. If the commits are easy to disentangle I'll make it separate.
- Yes, though I think the R gbm package also just supports standard right censoring (though with a 2-label format). Do you have plans to add complex censoring soon?
- Good point about the sorting being an issue with wrapper code. Let me check the performance of just doing a sort. If it's cheap we could just fall back to a sort alg whenever we encounter an unsorted data array. Then presorting is just an optimization if people want to use it.
- I planned to also include Eron's later for another project that had a bunch of ties. We could just start with though, and not offer breslow's at all.

slundberg commented:
Pulling the cox changes out was not easy, so I just squashed things to clean it all up.

- The cox loss now builds an argsort of the data and does not assume a sorted dataset
- I cleaned up the whitespace issues.
- Is there anywhere where RAM is currently queried in the codebase? Since it is an OS dependent check I wanted to be consistent with any other cases (which I have not found by looking).

khotilov commented on python-package/xgboost/core.py:
```diff
@@ -1056,7 +1063,12 @@ def predict(self, data, output_margin=False, ntree_limit=0, pred_leaf=False,
+            if interaction_contribs:
+                s = int(np.sqrt(ncol))
+                assert s*s == ncol, "XGBoost returned a non-square interaction matrix!"
+                preds = preds.reshape(nrow, int(np.sqrt(ncol)), int(np.sqrt(ncol)))
```
This doesn't work for multiclass.

slundberg replied:
good point. should the regular labels also break things out by class as a dimension? It would be a breaking change, but might make more sense

khotilov replied:
If the order of indexing would be well documented, maybe with some brief examples, then breaking things by class is not really necessary in python, as reshaping is quite simple and common operation for numpy arrays. The "discoverability" would be better though if things come pre-shaped... So I don't have much of an opinion towards either way here.

slundberg replied:
I went ahead and pre-shaped them for both interactions and regular contributions...since digging into the docs is less pleasant that having it be obvious.

khotilov commented on python-package/xgboost/core.py:
```diff
@@ -1040,6 +1045,8 @@ def predict(self, data, output_margin=False, ntree_limit=0, pred_leaf=False,
         if approx_contribs:
             option_mask |= 0x08
+        if interaction_contribs:
+            option_mask |= 0x16
```
x16 is 0b10110
Was the intent to turn on, in addition, the flags for pred_contribs and **pred_leaf**?

slundberg replied:
Good catch. No I was thinking decimal and writing hexadecimal

khotilov commented on doc/parameter.md:
```diff
@@ -170,6 +170,7 @@ Specify the learning task and the corresponding learning objective. The objectiv
     they can only be used when the entire training session uses the same dataset
   - "count:poisson" --poisson regression for count data, output mean of poisson distribution
     - max_delta_step is set to 0.7 by default in poisson regression (used to safeguard optimization)
+  - "survival:cox" --Cox regression for censored survival time data (negative values are considered censored)
```
For *right* censored...
It would also be worth mentioning that censoring is encoded by times with negative values and that predictions are returned on the hazard ratio scale (i.e., as `HR = exp(marginal_prediction)` in the proportional hazard function `h(t) = h0(t) * HR`).

khotilov commented:
> Is there anywhere where RAM is currently queried in the codebase? Since it is an OS dependent check I wanted to be consistent with any other cases (which I have not found by looking)

No, there's nothing currently. I think it would be useful to have the getMemorySize() code from here http://nadeausoftware.com/articles/2012/09/c_c_tip_how_get_physical_memory_size_system in src/common/ . I've checked that it works on at least windows and linux. The license is rather permissive from my point of view.

> I agree that a sparse matrix format would make sense. To really make that work though we would need to integrate a sparse representation into the original feature attribution code, since the interaction code builds on it.

Unless I'm missing something, I don't think it would affect the original feature attribution code. You just need to redefine the indexing of out_contribs in PredictInteractionContributions in a certain way. E.g., it's easy in a case with a set limit on the number of the largest contributions, say Nx, then there would be Nclass * Nx * 2 per row (each feature has two columns: its index and its contribution).

> Yes, though I think the R gbm package also just supports standard right censoring (though with a 2-label format). Do you have plans to add complex censoring soon?

gbm3 has right censoring and interval censoring (as counting process), and it looks like they now allow to have strata... My main reservation for not adding Cox models to xgboost was its proportionality assumption, and somehow I was rarely seeing that in data. ranger and randomForestSRC have fully non-parametric survival trees with vector leaves. Xgboost tree class has an unutilized ability to have vector leaves, However, I didn't give much thought to how survival trees could be done with boosting.

> I planned to also include Eron's later for another project that had a bunch of ties. We could just start with though, and not offer breslow's at all.

Keeping Breslow as an option might be useful.

BTW, what is the meaning of the values of interaction contributions? I don't see them summing up to predictions. Are they only comparable between each-other, or somehow also to the original SHAP contributions?

slundberg commented on include/xgboost/data.h:
```diff
@@ -96,6 +110,10 @@ struct MetaInfo {
+
+private:
+  /*! \brief argsort of labels */
+  mutable std::vector<size_t> label_order_cache;
```
done

slundberg commented on include/xgboost/tree_model.h:
```diff
@@ -501,13 +501,19 @@ class RegTree: public TreeModel<bst_float, RTreeNodeStat> {
-                       bst_float parent_one_fraction, int parent_feature_index) const;
+                       bst_float parent_one_fraction, int parent_feature_index,
+                       int condition, unsigned condition_feature,
+                       bst_float condition_fraction) const;
```
done

slundberg commented on src/predictor/cpu_predictor.cc:
```diff
@@ -279,6 +280,41 @@ class CPUPredictor : public Predictor {
+        for (size_t k = 0; k < ncolumns + 1; ++k) {
+          for (int l = 0; l < model.param.num_output_group; ++l) {
+            const unsigned cind = j * crow_chunk + k * ngroup + l;
+            const unsigned oind = j * row_chunk + i * mrow_chunk + k * ngroup + l;
```
Great point. I've made the ordering consistent.

As for saving time by only computing half...that might be possible, but the `PredictContribution` function does the real work, and it did not seem easy to save much time there without major changes.

khotilov replied:
Ah, I see... Yes, it looks like that wouldn't bring any appreciable time saving.

slundberg commented on tests/python/test_basic.py:
```diff
@@ -306,3 +309,157 @@ def test_fn(max_depth, num_rounds):
     assert out[0, 2] == 0.25
+
+
+    def parse_model(model):
```
done.

slundberg commented:
Your question about the meaning of the interaction values is a good one (and expecting them to sum up to something). Since the contributions come from toggling features on and off (condition on them or integrate them out) we have 2^M subsets of features. Each subset can be assigned a value m(f,S), and the output of the model (f) for some subset of features turned on can then be computed as the sum of m(f,T) over all subsets T in S (this is the mobius transform of the function). For an additive model m(f,S) will be 0 for all sets S except the singleton sets. This means we know the expected output of the model by just adding up m(f,{i}) for each feature i that is turned on.

For non-linear models m(f,S) might be non-zero for any S. The SHAP values result from evenly dividing all the m(f,S) values among the features in S. The interaction values result from dividing all the m(f,S) values up among the pairs of features in S. This leaves the m(f,{i}) unaccounted for which is why they don't sum up to match the original SHAP values.

To make m(f,{i}) easier to get at I set the diagonal of the interaction matrix to m(f,{i}). This puts the additive effects along the diagonal and the interaction effects off the diagonal. So now if you sum the matrix columns (or rows) you get the SHAP values.

slundberg commented:
> Unless I'm missing something, I don't think it would affect the original feature attribution code. You just need to redefine the indexing of out_contribs in PredictInteractionContributions in a certain way. E.g., it's easy in a case with a set limit on the number of the largest contributions, say Nx, then there would be Nclass * Nx * 2 per row (each feature has two columns: its index and its contribution).

You are right, we can return a sparse matrix. But without updating the underlying `PredictContribution` code we will still be creating dense rows of the matrix before throwing most of them away. The real trick would be inside the recursive `TreeSHAP` method we would need to use some kind of sparse random access data format such as a hash table. But there might be some significant speed tradeoffs there. I think sparse would be great eventually, but will take time to get right I think.

slundberg commented:
> gbm3 has right censoring and interval censoring (as counting process), and it looks like they now allow to have strata... My main reservation for not adding Cox models to xgboost was its proportionality assumption, and somehow I was rarely seeing that in data. ranger and randomForestSRC have fully non-parametric survival trees with vector leaves. Xgboost tree class has an unutilized ability to have vector leaves, However, I didn't give much thought to how survival trees could be done with boosting.

I am not enough of an expert on survival models to propose the ultimately best one for tree models, but given the broad use of cox, being able to use it in XGBoost (and then explain it with SHAP contribs) is valuable to those who would have used cox anyway with a different package (linear or otherwise).

slundberg commented:
> No, there's nothing currently. I think it would be useful to have the getMemorySize() code from here http://nadeausoftware.com/articles/2012/09/c_c_tip_how_get_physical_memory_size_system in src/common/ . I've checked that it works on at least windows and linux. The license is rather permissive from my point of view.

For right now I just hard coded a 1GB limit on the output matrix (with a print statement). Which is probably plenty since it would take a long time to compute that much, and it is simple for a user to run the code in batches if they want more.

I didn't use the the getMemorySize() code because from what I could tell the CC license is not technically compatible with the Apache (used by XGBoost). I suppose one could claim it is a "snippet" or something else, but I didn't feel qualified to vouch for that.

khotilov commented:
Thanks! I was trying to see whether the things look meaningful with a simple simulated data in https://gist.github.com/khotilov/f665813b766fc8a16a9e5c11e98aad89 , and to some degree they do. Would you be able to check whether I was comparing right things to each other in there? I'm kind of surprised that subtracting *half* of the interaction contributions' sums results in predictions that are very close to the predictions of a no-interactions linear model. But perhaps in this specific case they just happen to meet "half-way".

WRT sparse output, my primary concern was about the memory consumption of the output. Since the output is filled row-by-row, we would still have a temporary dense matrix built for each row and group, but then only a limited set of the largest contributions would be selected from it into the output. I think it would be a useful option to add, if not now then later.

WRT memory check: I would prefer either having it done properly or not done at all. If one would want to run overnight on a machine with a lot of memory, they should be able to, without having to deal with batches just because of an ad-hoc hardcoded limit. I'm curious: how long in ballpark does it take to fill a 1G output on a typical machine?

As for the [getMemorySize() code](http://nadeausoftware.com/articles/2012/09/c_c_tip_how_get_physical_memory_size_system) license,  I've seen that the Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0) license was [considered as compatible with Apache](https://www.apache.org/legal/resolved.html#cc-sa), and I assumed that the Attribution 3.0 Unported (CC BY 3.0) would be less restrictive than the CC BY-SA 3.0. But my assumptions could well be wrong, and maybe we should check with @pommedeterresautee ?  In any case, the memory check would not be a crucial feature to have, but just a nice safety guard, and it could always be considered later.

khotilov commented on src/learner.cc:
```diff
@@ -443,9 +443,12 @@ class LearnerImpl : public Learner {
+               bool interaction_contribs) const override {
+    if (pred_contribs && !interaction_contribs) {
       gbm_->PredictContribution(data, out_preds, ntree_limit, approx_contribs);
+    } else if (pred_contribs && interaction_contribs) {
```
Is it really necessary for interactions to have both pred_contribs and interaction_contribs on?

slundberg replied:
Good point. I'll just let interaction_contribs stand alone

slundberg commented:
Running through a basic example is a good idea, so I ran through the gist example and then made my own version of it. Specifically I made two simple models that we know the exact answer to and then checked that everything makes sense. It's at: https://github.com/slundberg/shap/blob/master/notebooks/Basic%20SHAP%20Interaction%20Value%20Example%20in%20XGBoost.ipynb As for the subtractions you were doing they now produce 0, before they should have produced the bias term (see below).

In the process of making `interaction_contribs` stand alone I realized it would be better named `pred_interactions` for consistency. I also increased the dimension of the interaction matrix to ncol+1 so it can include the bias term. This makes it consistent with `pred_contribs` and means that `shap_interaction_values.sum((1,2))` will equal the marginal prediction of the model.

If you are okay with make the sparse memory option and the memory limit separate issues that might be cleaner. Sounds like you could be right about the license, but it still might be good to keep that as a separate issue in case it comes up again. I am not ready to jump into the sparse stuff quite yet, but I would be happy to open another PR for the mem-check code. For right now I removed the current memcheck. It takes the same time to fill 1GB as it would with pred_contribs (it is only slower by a factor of ~2x per gigabyte). The run time depends on the number and depth of the trees, and number of features. I could imagine filling out 1GB with a simple model of many features.

slundberg commented:
Looks like the remaining Travis issue is about a missing package related to R. So I think it's unrelated to this PR

khotilov commented:
I'd personally rather not clump the interactions and the main effects together in the output of pred_interactions (taking even more space, awkward to extend to sparse interactions output in future), but ok...

And I still think that the off-diagonal interactions are a factor of two larger than they have to be WRT to the diagonal. In your example, the double-difference is an interaction term itself, but you are double-counting it by assigning it to both symmetric off-diagonal elements without dividing in half. See the updated and corrected gist at https://gist.github.com/khotilov/f665813b766fc8a16a9e5c11e98aad89

slundberg commented:
You are right: (Thank you!)
```
intr2 = intr.copy()
for i in range(intr2.shape[0]):
    for j in range(intr2.shape[1]):
        for k in range(j, intr2.shape[2]):
            intr2[i,k,j] = 0
half_intr = cont - intr2.sum(2)
np.corrcoef(half_intr.sum(1), y)[0,1]**2
```
Gives the same R-squared as a linear model. So this means `inter[0][np.triu_indices(inter.shape[1])].sum() == pred[0]` should hold, rather than a sum over the entire matrix. 

How would you prefer to return the main effects? Or would you prefer to let people compute them?

khotilov commented:
That tweak does seem to give correct diagonal sums, correct upper triangular sums, and correct upper triangular *row* sums (after quite more of work to do that). 

However, it still seems to produce distorted (towards larger interactions scale) individual interactions and the main terms for interacting features. E.g., after running my gist example above I see the following:
```python
true_interactions[0] # this was the result befor the today's change
Out[27]: 
array([[ 1.003,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  1.779, -0.521,  0.   ,  0.   ,  0.   ],
       [ 0.   , -0.521,  0.267,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]])

intr[0] # this is the current result
Out[28]: 
array([[ 1.005,  0.007, -0.009, -0.   ,  0.   ,  0.   ],
       [ 0.007,  1.762, -0.504, -0.   ,  0.   ,  0.   ],
       [-0.009, -0.504, -0.254,  0.   ,  0.   ,  0.   ],
       [-0.   , -0.   ,  0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]], dtype=float32)

gt_intr[0] # this is my ground truth calculation based on main and interactions terms values
Out[29]: 
array([[ 1.003,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  1.51 , -0.251,  0.   ,  0.   ,  0.   ],
       [ 0.   , -0.251, -0.003,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]])
```
I.e., I don't see the intr[:, 2, 2] being small, as I would expect it to be (from the formula with zero main effect from it in the data with zero-centered features), and the interactions are twice larger than I would expect. What is not correct with my assumption that this matrix should map to the original formula terms with half-split attribution of interactions in zero-centered data? Here, correspondence to the individual original terms is the ultimate check for me, rather than the aggregations.  Am I missing something?

khotilov commented:
Also, when I simply switch the order of the two interacting features 1 & 2 in my gist, it changes the results for the main terms of interactions a lot with your current version:
```python
true_interactions[0] # looks as expected
Out[37]: 
array([[ 1.003,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.267, -0.521,  0.   ,  0.   ,  0.   ],
       [ 0.   , -0.521,  1.779,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]])

intr[0] # looks fishy - really different main terms
Out[38]: 
array([[ 1.005, -0.009,  0.007, -0.   ,  0.   ,  0.   ],
       [-0.009,  0.25 , -0.504,  0.   ,  0.   ,  0.   ],
       [ 0.007, -0.504,  1.259, -0.   ,  0.   ,  0.   ],
       [-0.   ,  0.   , -0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]], dtype=float32)

gt_intr[0] # looks as expected
Out[39]: 
array([[ 1.003,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   , -0.003, -0.251,  0.   ,  0.   ,  0.   ],
       [ 0.   , -0.251,  1.51 ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   , -0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]])
```

As for the combining the main effects in the return of interactions, I'm ok with that as long as it provides a clear and meaningful interface that is easy to use for smaller number of features. The output interpretation and handling for sparse features would need to be different and more awkward anyway.

slundberg commented:
You are right switching features should not change the values of anything. When I just subtracted half the matrix that didn't treat everything symmetrically. After thinking it over I decided it would make the most sense (like you mentioned earlier) to multiply the interaction values by 0.5 before storing them in the off diagonals. Then the summation of the whole matrix still makes sense, and we can just subtract all of them from the SHAP values to get the main effects.

To double check that features with no main effects get a 0 attribution on the diagonal when the features are mean centered I updated https://github.com/slundberg/shap/blob/master/notebooks/Basic%20SHAP%20Interaction%20Value%20Example%20in%20XGBoost.ipynb with a simple example and it works.

This is new result with your gist:
```
> inter[0]
array([[ 1.004,  0.003, -0.004, -0.   ,  0.   ,  0.   ],
       [ 0.003,  1.507, -0.252, -0.   ,  0.   ,  0.   ],
       [-0.004, -0.252,  0.002,  0.   ,  0.   ,  0.   ],
       [-0.   , -0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ],
       [ 0.   ,  0.   ,  0.   ,  0.   ,  0.   , -0.002]], dtype=float32)
```

Sorry for the bugs when I added this whole main effect idea.

khotilov commented:
There's nothing to be sorry about. This whole thing is quite a mind-twister. And thanks for implementing it and making it work in a meaningful way!
